### PR TITLE
[PD] add parameter enabling logic for pad/pocket

### PR DESCRIPTION
**Pull request by donovaly**
- currently all parameters are always enabled and thus can lead to errors like the one reported here:
https://forum.freecadweb.org/viewtopic.php?f=3&t=70266

### CODE REVIEW

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198">pipeline status</a> for feature branch [PR_7198](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198). Pipeline [589587238](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/589587238) was triggered at [40d04a1](https://github.com/FreeCAD/FreeCAD/pull/7198/commits/40d04a1f11fde9d7188679415e595f380398d56a). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198">pipeline status</a> for feature branch [PR_7198](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198). Pipeline [589587543](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/589587543) was triggered at [6ae074d](https://github.com/FreeCAD/FreeCAD/pull/7198/commits/6ae074dd73ef65c415566cba7cafad140ebe2cae). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198">pipeline status</a> for feature branch [PR_7198](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198). Pipeline [589587984](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/589587984) was triggered at [cf5eaa2](https://github.com/FreeCAD/FreeCAD/pull/7198/commits/cf5eaa2b794b9e71883193fcc343496d6a63aedc). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198">pipeline status</a> for feature branch [PR_7198](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198). Pipeline [589636690](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/589636690) was triggered at [640137f](https://github.com/FreeCAD/FreeCAD/pull/7198/commits/640137f52349d0187b563f37b39cbff0e4778107). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/PartDesign/App/FeaturePad.cpp [hunk 1]
```diff
@@ -238,6 +238,52 @@ App::DocumentObjectExecReturn *Pad::execute()
            this->Shape.setValue(getSolid(prism));
         }
 
+        // eventually disable some settings that aer not valid on the current setup
```
Comment (AjinkyaDahale) [inline src/Mod/PartDesign/App/FeaturePad.cpp:241]: settings that _are_ not valid

Comment (donovaly) [reply src/Mod/PartDesign/App/FeaturePad.cpp]: Thanks, fixed now.

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198">pipeline status</a> for feature branch [PR_7198](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198). Pipeline [589803458](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/589803458) was triggered at [4362684](https://github.com/FreeCAD/FreeCAD/pull/7198/commits/436268455e5cdf9e01f21d568487c2253b6229e6). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198">pipeline status</a> for feature branch [PR_7198](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198). Pipeline [589804165](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/589804165) was triggered at [446347f](https://github.com/FreeCAD/FreeCAD/pull/7198/commits/446347ff1f0c86cc615c5327d17d1307eb8c5421). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by wwmayer** [comment]
First of all, the new code should go into a separate function.

Second, It's again a lot of code duplication and this can be easily handled by a single function.

The only difference between the two blocks is the UpToLast vs ThroughAll methods but it doesn't harm to put this into a common function because for a pad the ThroughAll will never been handled and vice versa for a pocket the UpToLast will never be handled.

**Comment by donovaly** [comment]
I am also not happy with this but failed to put it into FeatureExtrude. The only thing that worked was to setup a new file wit a single new function that I can call from within Pad or packet.
Should I do this?

**Comment by wwmayer** [comment]
Add a protected method updateProperties() to FeatureExtrude and implement it like this:

```
void FeatureExtrude::updateProperties(const std::string& method)
{
    // eventually disable some settings that are not valid on the current setup
    // disable everything unless we are sure we don't need it
    bool isLengthEnabled = false;
    bool isLength2Enabled = false;
    bool isOffsetEnabled = false;
    bool isMidplaneEnabled = false;
    bool isReversedEnabled = false;
    bool isUpToFaceEnabled = false;
    bool isTaperVisible = false;
    bool isTaper2Visible = false;
    if (method == "Length") {
        isLengthEnabled = true;
        isTaperVisible = true;
        isMidplaneEnabled = true;
        isReversedEnabled = !Midplane.getValue();
    }
    else if (method == "UpToLast") {
        isOffsetEnabled = true;
        isReversedEnabled = true;
    }
    else if (method == "ThroughAll") {
        isMidplaneEnabled = true;
        isReversedEnabled = !Midplane.getValue();
    }
    else if (method == "UpToFirst") {
        isOffsetEnabled = true;
        isReversedEnabled = true;
    }
    else if (method == "UpToFace") {
        isOffsetEnabled = true;
        isReversedEnabled = true;
        isUpToFaceEnabled = true;
    }
    else if (method == "TwoLengths") {
        isLengthEnabled = true;
        isLength2Enabled = true;
        isTaperVisible = true;
        isTaper2Visible = true;
        isReversedEnabled = true;
    }
    
    Length.setReadOnly(!isLengthEnabled);
    AlongSketchNormal.setReadOnly(!isLengthEnabled);
    Length2.setReadOnly(!isLength2Enabled);
    Offset.setReadOnly(!isOffsetEnabled);
    TaperAngle.setReadOnly(!isTaperVisible);
    TaperAngle2.setReadOnly(!isTaper2Visible);
    Midplane.setReadOnly(!isMidplaneEnabled);
    Reversed.setReadOnly(!isReversedEnabled);
    UpToFace.setReadOnly(!isUpToFaceEnabled);
}

```

In Pad::execute() or Pocket::execute() write `updateProperties(method);`

**Comment by donovaly** [comment]
> Add a protected method updateProperties() to FeatureExtrude and implement it like this:

That was exactly what I tried but it did not compile complaining about the Length property unresolved. I thought this is because This is set in FeaturePad and then it cannot be set in FeatureExtrude. But no, I just had to clean the PD target in MSVC and then everything works fine. Argh!

However done now.

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198">pipeline status</a> for feature branch [PR_7198](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7198). Pipeline [589913623](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/589913623) was triggered at [bc78a97](https://github.com/FreeCAD/FreeCAD/pull/7198/commits/bc78a97fce431e16d4e99ce396d80d1ccd18324a). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Pull request merged**